apps/examples/araplayer/HLSParser.cpp

```diff
@@ -0,0 +1,281 @@
+			++data;
+		}
+
+		if (*data == 0) {
+			break;
+		}
+
+		if (strncmp(TARGET_DURATION, data, mTargetDurationLen) == 0) {
+			int target_duration = 0;
+			data += mTargetDurationLen + 1;
+			while (*data >= '0' && *data <= '9') {
+				target_duration = target_duration * 10 + (*data - '0');
+				++data;
+			}
+			playlist->targetDuration = target_duration;
+		} else if (strncmp(BEGIN_PLAYLIST_URL, data, mBeginPlaylistUrlLen) == 0) {
+			if (playlist->type == play_list_type_t::NONE) {
+				playlist->type = play_list_type_t::MASTER;
+				mMapMasterPlaylist.insert(make_pair(std::string(url), playlist));
+			}
+
+			while (strncmp(BANDWIDTH, data, mBandwidthLen) != 0) {
+				++data;
+			}
+			data += mBandwidthLen + 1;
+
+			int bandwidth = 0;
+			while (*data >= '0' && *data <= '9') {
+				bandwidth = bandwidth * 10 + (*data - '0');
+				++data;
+			}
+
+			while (*data != '\n') {
+				++data;
+			}
+			++data;
+
+			idx = 0;
+			while (*data != '\r' && *data != '\n') {
+				segUrl[idx++] = *data++;
+			}
+
+			segUrl[idx] = 0;
+
+			hls_play_list_t *mediaPlaylist = new hls_play_list_t;
+			mediaPlaylist->url = std::string(segUrl);
+			mediaPlaylist->type = play_list_type_t::MEDIA;
+			mediaPlaylist->bandwidth = bandwidth;
+			mediaPlaylist->parent = playlist;
+
+			playlist->vecUrl.push_back(mediaPlaylist->url);
+			playlist->mapMediaPlaylist.insert(make_pair(mediaPlaylist->url, mediaPlaylist));
+		} else if (strncmp(BEGIN_SEGMENT_URL, data, mBeginSegmentUrlLen) == 0) {
+			if (playlist->type == play_list_type_t::NONE) {
+				playlist->type = play_list_type_t::SINGLE;
+				if (playlist->parent == NULL) {
+					DEBUG_TRACE(LOG_DEBUG, "Single Media Playlist [%s]", url);
+					mMapMasterPlaylist.insert(make_pair(std::string(url), playlist));
+				}
+			}
+
+			data += mBeginSegmentUrlLen + 1;
+			int point = 0;
+			double duration = 0;
+			while (*data != ',') {
+				duration = duration * 10 + (*data - '0');
+				point *= 10;
+				if (*data == '.') {
+					point = 1;
+				}
+				++data;
+			}
+
+			if (point) {
+				duration /= point;
+			}
+
+			++data;
+			while (*data == '\r' || *data == '\n') {
+				++data;
+			}
+
+			idx = 0;
+			while (*data != '\r' && *data != '\n') {
+				segUrl[idx++] = *data++;
+			}
+
+			segUrl[idx] = 0;
+			playlist->vecUrl.push_back(std::string(segUrl));
+		} else {
+			++data;
+		}
+	}
+
+	playlist->parsingCnt++;
+	if (mCallbackObject != NULL && mCBGetPlaylistInfo != NULL) {
+		if (isFirstParsedMediaPlaylist(playlist)) {
+			pInfo.playlistUrl = playlist->parent->url;
+		} else {
+			pInfo.playlistUrl = url;
+		}
+
+		getMediaPlaylistInfo(playlist, pInfo);
+		mCBGetPlaylistInfo(mCallbackObject, pInfo.playlistUrl.c_str(), pInfo);
+	}
+
+	return 1;
+}
+
+int HLSParser::getPlaylistInfo(const char *url, int bandwidth, has_segment_info_t *outSegInfo)
+{
+	int ret = 0;
+	PlaylistIter iter = mMapMasterPlaylist.find(url);
+	if (iter == mMapMasterPlaylist.end() || iter->second->parsingCnt <= 0) {
```

Comment:
please reverse flow. variable check -> function check is always cheaper.

Reply:
@tool3210 could you explain?